src/SpinWave/SWUtil.jl

```diff
@@ -0,0 +1,236 @@
+    energy_ϵ   :: Float64 # energy epsilon in the diagonalization. Set to add to diagonal elements of the spin-wave Hamiltonian for cholesky decompostion
+    energy_tol :: Float64 # energy tolerance for maximal imaginary part of spin-wave energies
+end
+
+
+"""
+    dipole_to_angles
+
+convert the dipole expectation values from the Cartesian frame to the spherical frame
+"""
+function dipole_to_angles(dipoles :: AbstractVector{Float64})
+    r = norm(dipoles)
+    @assert r > 1e-7
+    θ = acos(dipoles[3] / r)
+    @assert isfinite(θ)
+    ϕ = atan(dipoles[2], dipoles[1])
+    @assert isfinite(ϕ)
+    (ϕ < 0.0) && (ϕ += 2.0 * π)
+    return θ, ϕ
+end
+
+"""
+    generate_local_sun_gens
+
+Compute SU(N) generators in the local reference frame.
+"""
+function generate_local_sun_gens(sys :: System)
+    Nₘ, N = length(sys.dipoles), sys.Ns[1] # number of magnetic atoms and dimension of Hilbert space
+    if sys.mode == :SUN
+        s_mat = spin_matrices(N)
+
+        s̃_mat = Array{ComplexF64, 4}(undef, N, N, 3, Nₘ)
+        T̃_mat = Array{ComplexF64, 3}(undef, N, N, Nₘ)
+
+        U_mat = Matrix{ComplexF64}(undef, N, N)
+
+        for site = 1:Nₘ
+            U_mat[:, 1] = sys.coherents[1, 1, 1, site]
+            U_mat[:, 2:N] = nullspace(U_mat[:, 1]')
+            @assert isapprox(U_mat * U_mat', I) "rotation matrix from (global frame to local frame) not unitary"
+            for μ = 1:3
+                s̃_mat[:, :, μ, site] = Hermitian(U_mat' * s_mat[μ] * U_mat)
+            end
+            T̃_mat[:, :, site] = Hermitian(U_mat' * sys.interactions_union[site].aniso.matrep * U_mat)
+        end
+
+    elseif sys.mode == :dipole
+        s_mat_2 = spin_matrices(2)
+        s_mat_N = spin_matrices(N)
+        
+        s̃_mat = Array{ComplexF64, 4}(undef, 2, 2, 3, Nₘ)
+        T̃_mat = Array{ComplexF64, 3}(undef, 2, 2, Nₘ)
+
+        U_mat_2 = Matrix{ComplexF64}(undef, 2, 2)
+        U_mat_N = Matrix{ComplexF64}(undef, N, N)
+
+        for site = 1:Nₘ
+            θ, ϕ = dipole_to_angles(sys.dipoles[1, 1, 1, site])
+            U_mat_N[:] = exp(-1im * ϕ * s_mat_N[3]) * exp(-1im * θ * s_mat_N[2])
+            U_mat_2[:] = exp(-1im * ϕ * s_mat_2[3]) * exp(-1im * θ * s_mat_2[2])
+            @assert isapprox(U_mat_N * U_mat_N', I) "rotation matrix from (global frame to local frame) not unitary"
+            @assert isapprox(U_mat_2 * U_mat_2', I) "rotation matrix from (global frame to local frame) not unitary"
+            for μ = 1:3
+                s̃_mat[:, :, μ, site] = Hermitian(U_mat_2' * s_mat_2[μ] * U_mat_2)
+            end
+            T̃_mat[:, :, site] = Hermitian(U_mat_N' * sys.interactions_union[site].aniso.matrep * U_mat_N)[1:2, 1:2]
+        end
+    
+    # here I need help from Kipton and David to map back the solution, because in general this is difficult to implement.
+    elseif sys.mode == :large_S
+        error("unsupported")
+        # spin  = (N-1) / 2
+        # s_mat_2 = spin * spin_matrices(2)
```

Comment:
I'm happy to look at this. I wonder if we should simply 'abort' if certain complicated terms are present here (e.g. scalar biquadratic).

Reply:
Following our conversation, let's remove `:large_S` and instead provide instructions for people to perform their own anti-renormalization?